### PR TITLE
57 add lgmres algorithm

### PR DESCRIPTION
**Pull request by gusespinola**
Greetings! I think this `lgmres.m` script is somewhat ready for peer review, but I would like to receive your comments so I can improve it. I commented to @jccf19 that this version that I have been refactored is promising. However, it still has an important erratic behavior: the relative residual norms are non-monotonic -the expected behavior is to be non-increasing- when the tolerance is small -less than, e.g., 1e-6. Also, appropriate sanity chechks should be considered as well. Therefore, I open the pull request for your consideration, and look forward to your comments.

UPDATE: the non-monotonic behavior of the relative residual norms in `lgmres.m` seems to be already fixed.

### CODE REVIEW

File: src/lgmres.m [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        % Solve the least-squares problem
+        % s instead of m
+        Rs = HUpTri(1:s, 1:s);
+        gs = g(1:s);
+        minimizer = Rs \ gs;
+        % For LGMRES, we should now consider the approximation error
+        % separately as a variable: zName can be zCurrent,
+        % zCycle, zCurrentCycle, name is open to suggestions
+
+        W = zeros(n, s);
```
Comment (jhabriel) [inline src/lgmres.m:282]: Please try to document what is going on here, i.e., from L282 to L288

File: tests/test_lgmres.m [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+end
+
+% ----> Test for sanity checks
+
+function test_number_of_input_arguments()
+    % Test if error is raised when passing incorrect number of inputs.
+    % Error should be raised since 1 parameter is given
+    try
+        pd_gmres(ones(2, 2));
```
Comment (jhabriel) [inline tests/test_lgmres.m:42]:
```suggestion
        lgmres(ones(2, 2));
```

File: tests/test_lgmres.m [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+    % Compare with expected outputs
+    assertElementsAlmostEqual(x, [2; 3; 4]);
+    assert(flag == 1);
+    assertElementsAlmostEqual(relresvec, [1; 0]);
+    assert(time > 0 && time < 5);
+
+end
+
+function test_embree_three_by_three_toy_example()
```
Comment (jhabriel) [inline tests/test_lgmres.m:179]: This test be either delete it or adapted.

File: tests/test_lgmres.m [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            mvecExpected = [2; 2; 2; 3];
+        end
+        assertElementsAlmostEqual(relresvec, relresvecExpected);
+        assertEqual(mvec, mvecExpected);
+    end
+
+end
+
+function test_sherman_one()
+    % Test pd_gmres with sherman4 matrix
```
Comment (jhabriel) [inline tests/test_lgmres.m:226]:
```suggestion
    % Test lgmres with sherman1 matrix
```

File: tests/test_lgmres.m [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    [~, flag, relresvec, ~] = ...
+            lgmres(A, b, m, k, tol, maxit);
+
+    % We check if it has converged and the total sum of outer iterations
+    assertEqual(flag, 1);
+    assertEqual(size(relresvec, 1), 39);
+end
+
+function test_sherman_four()
+    % Test pd_gmres with sherman4 matrix
```
Comment (jhabriel) [inline tests/test_lgmres.m:249]:
```suggestion
    % Test lgmres with sherman4 matrix
```

File: tests/test_lgmres.m [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    A = Problem.A;
+    b = Problem.b;
+
+    % Setup LGMRES
+    m = 27;
+    k = 3;
+    tol = 1e-12;
+    maxit = 1000;
+
+    % Call PD-GMRES
```
Comment (jhabriel) [inline tests/test_lgmres.m:262]:
```suggestion
    % Call LGMRES
```

File: tests/test_lgmres.m [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
```
Comment (jhabriel) [inline tests/test_lgmres.m:239]:
```suggestion
    % Call LGMRES
```

File: tests/test_poisson.m [hunk 8]
```diff
@@ -84,8 +84,16 @@ function test_poisson_onedim_dir_bc()
...
+    % LGMRES
+    tol = 1e-9;
+    maxit = 100;
+    m = 3;
+    k = 1;
+    uLGMRES = lgmres(A, b, m, k, tol, maxit);
+
     % Assert whether the pd_gmres solution match the exact knonw solution
     assertElementsAlmostEqual(uPD_GMRES, uExact);
+    assertElementsAlmostEqual(uLGMRES, uExact);
```
Comment (jhabriel) [inline tests/test_poisson.m:96]: Excellent!

**Review by jhabriel — CHANGES_REQUESTED** [review]
Great job @gusespinola! 

I think we are almost there. Most of my suggestions are minor.

However, we should discuss whether it makes sense to add the sanity check regarding `k` and `m`. Let's discuss it in the next meeting. In the meantime, you can solve the other comments.

File: src/lgmres.m [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    %               Computational time in seconds.
+    %
+    %   References:
+    %   -----------
+    %
+    %   [1] Baker, A. H., Jessup, E. R., & Manteuffel, T. (2005). A technique
+    %   for accelerating the convergence of restarted GMRES. SIAM Journal on
+    %   Matrix Analysis and Applications, 26(4), 962-984.
+    %
+    %   [2] Nunez, R. C., Schaerer, C. E., & Bhaya, A. (2018). A
```
[the remark this answers is not in the document]
Comment (gusespinola) [reply src/lgmres.m:76]: Not anymore, just a residual from the PD-GMRES template.

File: src/lgmres.m [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+    restart = 1;
+    r0 = b - A * xInitial;
+    res(1, :) = norm(r0);
+    relresvec(1, :) = (norm(r0) / res(1, 1));
+    iter(1, :) = restart;
+
+    % Matrix with the history of approximation error vectors
+    zMat = zeros(n, k);
+
+    % while number_of_cycles <=k, according to 'mi_lgmres',
```
[the remark this answers is not in the document]
Comment (gusespinola) [reply src/lgmres.m:218]: `mi_lgmres` is a legacy LGMRES script. For what it's worth, the comment has been already solved and can be deleted.

File: src/lgmres.m (hunk 1 above)
Comment (gusespinola) [reply src/lgmres.m:282]: From Baker, fig. 1, lines 10 and 12, the n-by-s matrix W is created with the first m Arnoldi vectors and the k approximation error vectos, from newest to oldest. zCurrentCycle is the approximation error vector from the current outer iteration.

File: src/lgmres.m [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+    end
+
+    % Check whether b has the same number of rows as b
+    if rowsb ~= n
+        error("Dimension mismatch between matrix A and vector b.");
+    end
+
+    clear rowsb colsb;
+
+    % ----> We should evaluate special sanity checks for LGMRES here
```
[the remark this answers is not in the document]
Comment (gusespinola) [reply src/lgmres.m:143]: Last time we mentioned about writing a warning in the case the user inputs a number different than 1, 2 or 3. From my understanding, the Python implementation lacks sanity checks in the matter, but I share the script for a reading. https://github.com/scipy/scipy/blob/main/scipy/sparse/linalg/_isolve/lgmres.py

**Comment by gusespinola** [comment]
Greetings! I was performing some maintenance to the tests, and now I think this branch is ready to merge. Would you please check that everything is OK? Otherwise, I would like to know the adjustments to be made, if necessary. Thanks!

File: src/lgmres.m (hunk 11 above)
[the remark this answers is not in the document]
Comment (gusespinola) [reply src/lgmres.m]: I guess that, from our last meeting in the matter, all the discussed sanity checks were performed, but I look forward to some review.

**Comment by jhabriel** [comment]
Thanks for the effort @gusespinola. I will look into this, possibly not tomorrow but on Friday.

**Review by jhabriel — APPROVED** [review]
Great job @gusespinola. I will publish a new version of Krysbas when you merge this PR.

**Comment by gusespinola** [comment]
Thanks for checking this!

**Pull request merged**